### PR TITLE
Change getiter and iternext to not be pure.  Resolves #3425

### LINKED ISSUE #3425: Code hang bug on py36.  Maybe parfor related.
## Reporting a bug

- [x] I am using the latest released version of Numba (most recent is visible in
 the change log (https://github.com/numba/numba/blob/master/CHANGE_LOG).
- [x] I have included below a minimal working reproducer (if you are unsure how
 to write one see http://matthewrocklin.com/blog/work/2018/02/28/minimal-bug-reports).

The following hangs with threads usage max'ed out on osx, py3.6 on 9dfaa295cb43d43023add5254d5768651877787d and 0.40.0 released version.

Code:

```python
import numpy as np
from numba import njit, prange


@njit(parallel=True)
def core(volcano, anglebreaks, volcanoshadow):
    for i in prange(0, volcano.shape[0]):
        for j in range(0, volcano.shape[1]):
            #### Looping using the following line will deadlock
            #### Also, turning off parallel=True makes it work too
            for anglei in anglebreaks:
            ### Loop using the following lines will work
            # for m in range(len(anglebreaks)):
            #     anglei = anglebreaks[m]
                volcanoshadow[i, j] = volcanoshadow[i, j] - 1


def bench_rays(volcano):
    volcanoshadow = np.ones_like(volcano)
    angle = -90. / 180. * np.pi
    diffangle = 90. / 180. * np.pi
    nangles = 25
    anglebreaks = np.tan(np.linspace(angle, diffangle, nangles, endpoint=True))

    core(volcano, anglebreaks, volcanoshadow)
    return volcanoshadow


if __name__ == '__main__':
    shape = 87, 61
    volcano = np.ones(shape)
    print(volcano.shape)
    shade = bench_rays(volcano)
    print(shade)
```

The code works fine on py3.5.  Py3.7 not tested.

### PR DESCRIPTION
**Pull request by DrTodd13**
Resolves #3425.

is_pure in ir_utils.py is kind of scary.  It seems to be an opt-out of purity rather than an opt-in.  In #3425, a getiter and iternext were being hoisted and then with subsequent LLVM optimizations some block was reduced to a jmp to itself, ergo infinite loop and hang.  I did a scan of ir.py searching for other IR nodes that aren't pure but there's some I'm not sure about so @stuartarchibald could do a better job here.

### CODE REVIEW

File: numba/ir_utils.py [hunk 1]
```diff
@@ -637,19 +637,22 @@ def is_pure(rhs, lives, call_table):
...
-        if func_name not in call_table or call_table[func_name] == []:
-            return False
-        call_list = call_table[func_name]
-        if (call_list == [slice] or
-            call_list == ['log', numpy] or
-            call_list == ['empty', numpy]):
-            return True
-        for f in is_pure_extensions:
-            if f(rhs, lives, call_list):
+    if isinstance(rhs, ir.Expr)
```
Comment (seibert) [inline numba/ir_utils.py]: syntax error here.  Missing colon.

**Comment by seibert** [comment]
FYI: this is failing CI because there is a syntax error on line 640.

File: numba/ir_utils.py (hunk 1 above)
Comment (DrTodd13) [reply numba/ir_utils.py]: Ugh.  There was a merge or something and maybe accidental deletion there.  It certainly wasn't like this when I tested it.

**Comment by stuartarchibald** [comment]
@DrTodd13 I agree, `is_pure` is scary. Perhaps we should just invert the logic so that "pure" is opt-in opposed to assuming it is until told otherwise/surprises! As to the PR, the original triggering sample should probably be stripped down and added as a unit test in `test_parfors.py`.

Also, why is `np.log` explicitly singled out as pure, got merged here https://github.com/numba/numba/pull/3004?

I can also make other bad things happen having thought about the problem a bit, for example, I think this is hoisting a `build_set` when it shouldn't be?:
```python
from numba import njit, prange, threading_layer
import numpy as np

idx = 14

def test_impl(X):
    n = X.shape[0]
    acc = 0
    for i in prange(n):
        R = {1, 2, 3}
        R.add(i)
        tmp = 0
        if i == idx:
            print("loop", i)
        for x in R:
            tmp += x
            if i == idx:
                print(x)
        acc += tmp
    return acc

n = 128
np.random.seed(0)
X = np.random.ranf(n)
print(test_impl(X))
print(njit(test_impl)(X))
print(njit(parallel=True)(test_impl)(X))
print(threading_layer())
```
which does this:
```
loop 14
1
2
3
14
8890
loop 14
1
2
3
14
8890
loop 14
3.0
2.0
1.0
0.0
12.0
4.0
6.0
10.0
7.0
13.0
11.0
9.0
14.0
8.0
5.0
123790
workqueue
```

FWIW the parallel diagnostics tool picked up the problem in the original code, the instruction hoisting section has `getiter` and `iternext` listed for hoisting:
```
================================================================================
====== Parallel Accelerator Optimizing:  Function core, issue3425.py (5)  ======
================================================================================
---------------------------Looking for parallel loops---------------------------
Parallel for-loop #0: is produced from issue3425.py (7):
    user defined prange
 
Found 1 parallel loops.
--------------------------------------------------------------------------------


Parallel loop listing for file issue3425.py
------------------------------------------------------------------------------------|loop #ID
import numpy as np                                                                  | 
from numba import njit, prange                                                      | 
                                                                                    | 
                                                                                    | 
@njit(['void(float64[:,:], float64[:], float64[:,:])'], parallel=True)              | 
def core(volcano, anglebreaks, volcanoshadow):                                      | 
    for i in prange(0, volcano.shape[0]):-------------------------------------------| #0
        for j in range(0, volcano.shape[1]):                                        | 
            #### Looping using the following line will deadlock                     | 
            #### Also, turning off parallel=True makes it work too                  | 
            for anglei in anglebreaks:                                              | 
            ### Loop using the following lines will work                            | 
            #for m in range(len(anglebreaks)):                                      | 
                #anglei = anglebreaks[m]                                            | 
                volcanoshadow[i, j] = volcanoshadow[i, j] - 1                       | 
                                                                                    | 
                                                                                    | 
def bench_rays(volcano):                                                            | 
    volcanoshadow = np.ones_like(volcano)                                           | 
    angle = -90. / 180. * np.pi                                                     | 
    diffangle = 90. / 180. * np.pi                                                  | 
    nangles = 25                                                                    | 
    anglebreaks = np.tan(np.linspace(angle, diffangle, nangles, endpoint=True))     | 
                                                                                    | 
    core(volcano, anglebreaks, volcanoshadow)                                       | 
    return volcanoshadow                                                            | 
                                                                                    | 
                                                                                    | 
if __name__ == '__main__':                                                          | 
    shape = 87, 61                                                                  | 
    volcano = np.ones(shape)                                                        | 
    print(volcano.shape)                                                            | 
    shade = bench_rays(volcano)                                                     | 
    print(shade)                                                                    | 
--------------------------------- Fusing loops ---------------------------------
Attempting fusion of parallel loops (combines loops with similar properties)...
Following the attempted fusion of parallel for-loops there are 1 parallel for-
loop(s) (originating from loops labelled: #0).
--------------------------------------------------------------------------------
----------------------------- Before Optimisation ------------------------------
--------------------------------------------------------------------------------
------------------------------ After Optimisation ------------------------------
Parallel structure is already optimal.
--------------------------------------------------------------------------------
--------------------------------------------------------------------------------
 
---------------------------Loop invariant code motion---------------------------
Allocation hoisting:
No allocation hoisting found

Instruction hoisting:
loop #0 has the following hoisted:
  $46.2 = getiter(value=anglebreaks)
  $50.2 = iternext(value=$46.2)
  $anglei.15 = pair_first(value=$50.2)
  $50.4 = pair_second(value=$50.2)
  $24.1 = global(range: <class 'range'>)
  $const24.2 = const(int, 0)
  $24.4 = getattr(value=volcano, attr=shape)
  $const24.5 = const(int, 1)
  $24.6 = static_getitem(value=$24.4, index=1, index_var=$const24.5)
  $const52.7 = const(int, 1)
--------------------------------------------------------------------------------
```
following the patch:
```
================================================================================
====== Parallel Accelerator Optimizing:  Function core, issue3425.py (5)  ======
================================================================================
---------------------------Looking for parallel loops---------------------------
Parallel for-loop #0: is produced from issue3425.py (7):
    user defined prange
 
Found 1 parallel loops.
--------------------------------------------------------------------------------


Parallel loop listing for file issue3425.py
------------------------------------------------------------------------------------|loop #ID
import numpy as np                                                                  | 
from numba import njit, prange                                                      | 
                                                                                    | 
                                                                                    | 
@njit(['void(float64[:,:], float64[:], float64[:,:])'], parallel=True)              | 
def core(volcano, anglebreaks, volcanoshadow):                                      | 
    for i in prange(0, volcano.shape[0]):-------------------------------------------| #0
        for j in range(0, volcano.shape[1]):                                        | 
            #### Looping using the following line will deadlock                     | 
            #### Also, turning off parallel=True makes it work too                  | 
            for anglei in anglebreaks:                                              | 
            ### Loop using the following lines will work                            | 
            #for m in range(len(anglebreaks)):                                      | 
                #anglei = anglebreaks[m]                                            | 
                volcanoshadow[i, j] = volcanoshadow[i, j] - 1                       | 
                                                                                    | 
                                                                                    | 
def bench_rays(volcano):                                                            | 
    volcanoshadow = np.ones_like(volcano)                                           | 
    angle = -90. / 180. * np.pi                                                     | 
    diffangle = 90. / 180. * np.pi                                                  | 
    nangles = 25                                                                    | 
    anglebreaks = np.tan(np.linspace(angle, diffangle, nangles, endpoint=True))     | 
                                                                                    | 
    core(volcano, anglebreaks, volcanoshadow)                                       | 
    return volcanoshadow                                                            | 
                                                                                    | 
                                                                                    | 
if __name__ == '__main__':                                                          | 
    shape = 87, 61                                                                  | 
    volcano = np.ones(shape)                                                        | 
    print(volcano.shape)                                                            | 
    shade = bench_rays(volcano)                                                     | 
    print(shade)                                                                    | 
--------------------------------- Fusing loops ---------------------------------
Attempting fusion of parallel loops (combines loops with similar properties)...
Following the attempted fusion of parallel for-loops there are 1 parallel for-
loop(s) (originating from loops labelled: #0).
--------------------------------------------------------------------------------
----------------------------- Before Optimisation ------------------------------
--------------------------------------------------------------------------------
------------------------------ After Optimisation ------------------------------
Parallel structure is already optimal.
--------------------------------------------------------------------------------
--------------------------------------------------------------------------------
 
---------------------------Loop invariant code motion---------------------------
Allocation hoisting:
No allocation hoisting found

Instruction hoisting:
loop #0 has the following hoisted:
  $24.1 = global(range: <class 'range'>)
  $const24.2 = const(int, 0)
  $24.4 = getattr(value=volcano, attr=shape)
  $const24.5 = const(int, 1)
  $24.6 = static_getitem(value=$24.4, index=1, index_var=$const24.5)
  $const52.7 = const(int, 1)
--------------------------------------------------------------------------------
```

**Comment by ehsantn** [comment]
@DrTodd13 All nodes in your list are `pure` for the things Numba can handle I think. In general, things `getattr` can be overloaded and have side effects but Numba doesn't handle such things.

@stuartarchibald The set issue seems like a bug in hoisting other than `is_pure`.

**Comment by stuartarchibald** [comment]
@ehsantn yes, I agree. I was trying to find *something* to trip up `is_pure`, I think the hoist is invalid possibly because the `set.add` isn't handled correctly, so probably not an `is_pure` bug.

**Comment by DrTodd13** [comment]
Yes, I think build_set is pure but the hoisting algorithm must have a bug.  is_pure isn't a sufficient check if there is something that isn't SSA and is modified later in the loop.

**Comment by DrTodd13** [comment]
We are doing the SSA check just by looking at assignments so the call to the class member function isn't detected.

**Comment by DrTodd13** [comment]
Added some code to hoisting to look for things of the form:
```
a = obj.attr
a()
```
If such a call to a member function is found, presume that it could modify the object and thus add that object to the set of objects defined more than once so that that object won't get hoisted.

**Comment by DrTodd13** [comment]
@stuartarchibald I'm not sure the original case rises to the level of a test case.  If it did and there was a bug then what would the infinite loop due to travis testing?  I do like your second example as a test case.  What do you think?

**Comment by stuartarchibald** [comment]
Thanks for looking into this @DrTodd13, what you suggests RE SSA analysis fits with what I observed, thanks for fixing it.

WRT to original case being used as a test case, the only impact on CI of it being present and broken would be a timeout and then SIGKILL would be sent to the process. I do see your point though and I think with the present state of `parfors` we can leave the test case out as it adds nothing of value, I've checked manually and it works fine. However once https://github.com/numba/numba/pull/3160 is complete, the collected diagnostic information can be used to assert that e.g. hoisting is actually working in the manner expected on precisely the things we expect!

>  I do like your second example as a test case but how would we detect correctness? As it is, the output order is non-deterministic. What do you think?

I assume your are referring to printing to stdout or am I missing something? Can we just strip the print statements (they were just there to demonstrate the point) and assert the returned accumulator holds the correct value? Without the patch in https://github.com/numba/numba/pull/3455/commits/d0bae353465543a0a01c89474b024c1aea8f4a47, nonsense is returned.

**Comment by sklam** [comment]
@ehsantn, @DrTodd13 

>  things `getattr` can be overloaded and have side effects but Numba doesn't handle such things.

Numba supports that via the experimental Jitclass support and possible via extension types.  `is_pure` should consider the types for `getattr`; i.e. pure if `getattr(obj <: NumpyArray, attr)` 

https://github.com/numba/numba/issues/3459 may be related.

**Pull request merged**